### PR TITLE
Tpetra: Adding cmake support for making Tpetra allocate in shared space

### PR DESCRIPTION
**Pull request by csiefer2**
Issue: #12621
Stakeholder feedback: In progress
Tests: n/a

@mperego This (in theory) should accommodate the "allocate in CudaUVMSpace once the Kokkos option goes away."

This will also need #12620 to merge for it to actually be useful.

You will need to add the following to your Trilinos configure to make this work:
```
-D KokkosKernels_INST_MEMSPACE_CUDAUVMSPACE=ON
-D Tpetra_ALLOCATE_IN_SHARED_SPACE=ON

```

### CODE REVIEW

File: packages/tpetra/core/compat/Tpetra_KokkosCompat_ClassicNodeAPI_Wrapper.hpp [hunk 1]
```diff
@@ -69,16 +69,26 @@ class KokkosDeviceWrapperNode {
...
 #ifdef KOKKOS_ENABLE_SYCL
+#ifdef HAVE_TPETRA_SHARED_ALLOCS
+  typedef KokkosDeviceWrapperNode<::Kokkos::Experimental::SYCL, ::Kokkos::Experimental::SYCLSharedUSMSpace> KokkosSYCLWrapperNode;
+#else
   typedef KokkosDeviceWrapperNode<::Kokkos::Experimental::SYCL, ::Kokkos::Experimental::SYCLDeviceUSMSpace> KokkosSYCLWrapperNode;
 #endif
+#endif
 
 #ifdef KOKKOS_ENABLE_HIP
+  // NOTE: KokkosKernels does not currently suport HIPManagedSpace, but
```
Comment (jhux2) [inline packages/tpetra/core/compat/Tpetra_KokkosCompat_ClassicNodeAPI_Wrapper.hpp:80]: spelling

File: packages/tpetra/core/compat/Tpetra_KokkosCompat_ClassicNodeAPI_Wrapper.hpp [hunk 2]
```diff
@@ -69,16 +69,26 @@ class KokkosDeviceWrapperNode {
...
+  // if it did, we'd add HAVE_TPETRA_MANAGED_ALLOCS here
```
Comment (jhux2) [inline packages/tpetra/core/compat/Tpetra_KokkosCompat_ClassicNodeAPI_Wrapper.hpp:81]: would it be better to throw a cmake configure error, e.g., "HIP shared space not supported yet" if the user enables HIP and also requests the default allocation to be in shared space?

Comment (brian-kelley) [reply packages/tpetra/core/compat/Tpetra_KokkosCompat_ClassicNodeAPI_Wrapper.hpp:81]: ETI for ``<HIP, HIPManagedSpace>`` doesn't exist, but everything should still work with it, AFAIK. It just might take longer to build downstream packages. We will add the ETI option.

Comment (jhux2) [reply packages/tpetra/core/compat/Tpetra_KokkosCompat_ClassicNodeAPI_Wrapper.hpp]: So in other words, this option should be enabled for `HIPManagedSpace` as well?

**Review by jhux2 — APPROVED** [review]
lgtm

**Pull request merged**